### ISSUE TITLE
Adding some constraints to certain wells.

### ISSUE THREAD

**Issue by Jbonnett** [armstrong/armstrong#38]
It would be nice to have some sort of soft constraints on what can be in a well. For example here at the bay citizen we have our right rails which we are re-implementing in Armstrong Wells. We have 3 ads that should be in certain positions of every right rail well, unless there is a VERY good reason for them not to be.   Some sort of popup or warning, would help us enforce the biz constraints of having ads while still letting editorial folks change things in the right rail as they see fit.

Perhaps this could be implemented such that you would set up Well Templates and for each slot in the template you could set wither a type mismatch would raise a warning or a error.

**Comment by tswicegood** [comment]
I am +1 for having some sort of workflow engine built in to Armstrong in general.  I think this is just a specific use-case for some sort of onload system that is triggered as part of the page load on certain forms.

**Comment by Jbonnett** [comment]
Yeah I can see that. Solving it generically is going to be much more work, but probably also a much greater pay off.  We should probably keep the ticket around and change the label to something like .armstrong.apps.workflows or something like that, just so the example usecase is preserved.

**Comment by dmclain** [comment]
I'd have to know a bit more about what exactly @tswicegood means by a workflow engine before I know if I have an opinion. I'm against a general-purpose business rules enforcement layer unless that layer happens to be called "Python". Looking at the Bay Citizen site, I'm confused as to how exactly you're trying to use wells for the right rail. Are you trying to make it so editors can dynamically choose exactly which wells show up on the right side as well as the ads relative location between them?

**Comment by tswicegood** [comment]
@dmclain: I mean a system where you can say "run this code" (probably JS since it would be easiest done in the browser) on this form.  Think "add these three nodes when someone selects `right rail` as the well type".  Seems like that would be generic enough that you could add it to any form.

**Comment by Jbonnett** [comment]
Something like that is already implemented. We call it right rails and
modules, but its a pretty direct translation to wells.

To translate from baycitizen-ican to armstrong-ese a module could be any
hunk html/template, and might contain another well. On the site these
are generally in little boxes like the whats hot module or the pulse of the
bay. Its dynamically altered by editorial. Ie they might use the twitter
feed module to follow a certain feed right there on the page. The need for
constraints arises from the need for some items to remain the same without
being hard coded.

I am completely fine with having some spot to dump code to do that. Most
biz process languages end up approaching real language status, so we can
probably save ourselves some effort by just using python or js. I am
looking more for the tools to build this than a fully pre-built system.

**Comment by dmclain** [comment]
@tswicegood If you just want to add javascript to the admin form then I believe you can do that like so::

```
class CustomNodeInlineAdminForm(NodeInlineAdminForm):
    class Media:
        js = (
                'your/custom/javascript.js',
              )

class CustomNodeInline(BackboneInline):
    form = CustomNodeInlineAdminForm

class CustomWellAdmin(WellAdmin):
    inlines = [
        CustomNodeInline,
    ]

admin.site.register(models.Well, CustomWellAdmin)
```

@jbonnett You can do something akin to the above::

```
class CustomWellForm(ModelForm):
    sure = forms.BooleanField(label='Expert Mode')
    class Meta:
        model = Well

    def clean(self):
        cleaned_data = self.cleaned_data
        sure = cleaned_data.get("sure")
        if not sure:
            if cleaned_data.get("type") == 'right_rail':
                # your custom validation goes here

class CustomWellAdmin(WellAdmin):
    form = CustomWellForm
    inlines = [
        CustomNodeInline,
    ]

admin.site.register(models.Well, CustomWellAdmin)
```

I'm not actually sure how that works when it comes to the inlines which is what the original goal is, but it lets you do custom validation of armstrong objects.

**Comment by Jbonnett** [comment]
Yes, we do exactly that here for various reasons. If the answer is to punt, that's fine too. I will just move this over to our ticketing system.

**Comment by tswicegood** [comment]
The answer isn't to punt, it's that it's already doable.

We can't have specific one-off cases built into Armstrong, thus my request for something more generic.  I'd love to have it set so you can drop a JS file in a particular directory, named a particular way, and the admin will find it and add it, but right now I don't consider that a priority.
